### PR TITLE
📚 DOCS: Tutorial on Launching a WorkChain

### PR DESCRIPTION
**Pull request by flavianojs**
Introduces an example in the Tutorial section on how to launch a WorkChain through the REST API.
This example was adapted from the GSOC 2021 Report by Ninad Bhat.

### CODE REVIEW

File: docs/source/user_guide/tutorial.md [hunk 1]
```diff
@@ -1,17 +1,206 @@
 # Tutorial
 
-This page can contain a simple tutorial for your code.
+We will demonstrate the usage of the [aiida-restapi](https://github.com/aiidateam/aiida-restapi) package with a few examples.
 
-## What we want to achieve
+To follow this tutorial, you need to have [aiida-core](https://github.com/aiidateam/aiida-core) and [aiida-restapi](https://github.com/aiidateam/aiida-restapi) installed. The REST API server should be running. Check the [README.md](https://github.com/aiidateam/aiida-restapi/blob/master/README.md) file from the [aiida-restapi](https://github.com/aiidateam/aiida-restapi) package to know how.
 
-### Step 1
+You will also need a HTTP operator that allows you to send attachments (which will constitute what is called the `Request Body`). Here are two options:
```
Comment (ltalirz) [inline docs/source/user_guide/tutorial.md:7]:
```suggestion
You will also need a tool to make HTTP requests. Here are two options:
```

File: docs/source/user_guide/tutorial.md [hunk 2]
```diff
@@ -1,17 +1,206 @@
...
 
-Some text
+**Option 1: HTTPie**
```
Comment (ltalirz) [inline docs/source/user_guide/tutorial.md:9]:
```suggestion
**Option 1: HTTPie command line tool**
```

File: docs/source/user_guide/tutorial.md [hunk 3]
```diff
@@ -1,17 +1,206 @@
...
+Then you can execute the REST API call with
 
-Some text
+```console
+$ http localhost:8000/api/v4/endpoint < request_body
+```
+
+where `request_body` is the file containing the request body.
+
+**Option 2: [Resquests library](https://docs.python-requests.org/en/latest/) (all python approach)**
```
Comment (ltalirz) [inline docs/source/user_guide/tutorial.md:25]:
```suggestion
**Option 2: [`requests` python library](https://docs.python-requests.org/en/latest/) (all python approach)**
```

File: docs/source/user_guide/tutorial.md [hunk 4]
```diff
@@ -1,17 +1,206 @@
...
+
+## Launching a WorkChain
+
+Here, we show how the REST API can be used to submit a WorkChain process to add two integers.
+
+### Step 1: Post Code Object
+   
+Request URL: http://127.0.0.1:8000/nodes
+Request Body:
+```python
```
Comment (ltalirz) [inline docs/source/user_guide/tutorial.md:53]:
```suggestion
```json
```

This applies to all requests

File: docs/source/user_guide/tutorial.md [hunk 5]
```diff
@@ -1,17 +1,206 @@
...
+   },
+   "extras": {
+      "_aiida_hash": "72864a25c6ebf290a12d522b8819fb858788bf81730e934dc95ca7a1ff8cce5c"
+   }
+}
+```
+
+### Step 2: Post Integers
+
+We are creating two integers with values 6 and 3. Below is the example for posting orm.Int object with value 6.
```
Comment (ltalirz) [inline docs/source/user_guide/tutorial.md:90]:
```suggestion
We are creating two integers with values 6 and 3. Below is the example for posting an `orm.Int` object with value 6.
```

File: docs/source/user_guide/tutorial.md [hunk 6]
```diff
@@ -1,17 +1,206 @@
...
+   "attributes": {
+      "value": 6
+   },
+   "extras": {
+      "_aiida_hash": "d74556d1d8c2610f7ec28de9c18c57efaca06a0b478875e304f9fe05ac4213d6"
+   }
+}
+```
+
+Similarly, we created another orm.Int object of value 3. The retrieved UUID was “29ff3e11-5165-46db-84b0-d85620d0b972”.
```
Comment (ltalirz) [inline docs/source/user_guide/tutorial.md:125]:
```suggestion
Similarly, we created another `orm.Int` object of value 3. The retrieved UUID was “29ff3e11-5165-46db-84b0-d85620d0b972”.
```

File: docs/source/user_guide/tutorial.md [hunk 7]
```diff
@@ -1,17 +1,206 @@
...
+
+body = {
+   "node_type": "data.int.Int.|",
+   "attributes": {"value": 6},
+   "label": "test_Int_1"
+}
+
+response = requests.post(url, data=body)
+
+print( response.text )
```
Comment (ltalirz) [inline docs/source/user_guide/tutorial.md:42]: Maybe use `response.json()` instead

File: docs/source/user_guide/tutorial.md [hunk 8]
```diff
@@ -1,17 +1,206 @@
...
+      "prepend_text": "",
+      "process_label": "ArithmeticAddCalculation",
+      "process_state": "created",
+      "input_filename": "aiida.in",
+      "output_filename": "aiida.out"
+   },
+}
+```
+
+The process status can be retrieved from the response.
```
Comment (ltalirz) [inline docs/source/user_guide/tutorial.md:206]:
```suggestion
The process status is reported in `response.json()["attributes"]["process_state"]`.
```

**Review by ltalirz — CHANGES_REQUESTED** [review]
thanks a lot @flavianojs 

Just a couple of suggestions

File: docs/source/user_guide/tutorial.md [hunk 9]
```diff
@@ -1,17 +1,206 @@
...
+   "attributes": {
+      "is_local": false,
+      "remote_exec_path": "/bin/true"
+   },
+   "label": "test_code"
+}
+```
+
+Response Body:
+```python
```
Comment (flavianojs) [inline docs/source/user_guide/tutorial.md:66]:
```suggestion
```json
```

File: docs/source/user_guide/tutorial.md [hunk 10]
```diff
@@ -1,17 +1,206 @@
...
+Similarly, we created another orm.Int object of value 3. The retrieved UUID was “29ff3e11-5165-46db-84b0-d85620d0b972”.
+
+### Step 3: Posting Process
+
+After retrieving the UUIDs, the process is submitted at `/processes` endpoint as follows:
+
+   Request URL: http://127.0.0.1:8000/processes
+
+   Request Body:
+   ```python
```
Comment (flavianojs) [inline docs/source/user_guide/tutorial.md:134]:
```suggestion
   ```json 
```

File: docs/source/user_guide/tutorial.md [hunk 11]
```diff
@@ -1,17 +1,206 @@
...
+         "y.uuid": "29ff3e11-5165-46db-84b0-d85620d0b972",
+         "metadata": {
+               "description": "job submission with the adding processes."
+         }
+      }
+   }
+   ```
+
+   Response Body:
+   ```python
```
Comment (flavianojs) [inline docs/source/user_guide/tutorial.md:150]:
```suggestion
   ```json
```

File: docs/source/user_guide/tutorial.md [hunk 12]
```diff
@@ -1,17 +1,206 @@
...
+   ```
+
+### Step 4: Check status
+
+Once the process is submitted, its status can be checked:
+
+Request URL: http://127.0.0.1:8000/processes/64
+
+Response Body:
+```python
```
Comment (flavianojs) [inline docs/source/user_guide/tutorial.md:178]:
```suggestion
```json
```

File: docs/source/user_guide/tutorial.md (hunk 7 above)
Comment (flavianojs) [inline docs/source/user_guide/tutorial.md:42]:
```suggestion
print( response.json )
```

File: docs/source/user_guide/tutorial.md [hunk 13]
```diff
@@ -1,17 +1,206 @@
...
+```
+
+### Step 2: Post Integers
+
+We are creating two integers with values 6 and 3. Below is the example for posting orm.Int object with value 6.
+
+Request URL: http://127.0.0.1:8000/nodes
+
+Request Body:
+```python
```
Comment (flavianojs) [inline docs/source/user_guide/tutorial.md:95]:
```suggestion
```json
```

File: docs/source/user_guide/tutorial.md [hunk 14]
```diff
@@ -1,17 +1,206 @@
...
+{
+   "node_type": "data.int.Int.|",
+   "attributes": {"value": 6},
+   "label": "test_Int_1"
+}
+```
+
+Response Body:
+```python
```
Comment (flavianojs) [inline docs/source/user_guide/tutorial.md:104]:
```suggestion
```json
```

**Comment by ltalirz** [comment]
Thanks for the fixes @flavianojs , as well as for pointing out the authentication issue over slack.

The easiest steps for authentication are:

1. Go to http://127.0.0.1:8000/docs
2. Click on "Authorize"
3. Use username "johndoe@example.com" and password "secret"
4. Log in

You can now use the Swagger Javascript client to send requests to the API (which we should probably add as the first way of doing things in the tutorial).

What this does is it will add a `Authorization: Bearer eyJhbGciOiJIUzI1NiIs...` header to all requests.

You can also authenticate from the python API but let's not include that in the tutorial.

**Comment by ltalirz** [comment]
@flavianojs Are you still going to add the bit on the authentication or should I go ahead and merge this?

**Comment by flavianojs** [comment]
> @flavianojs Are you still going to add the bit on the authentication or should I go ahead and merge this?

Thanks, Leo. Give me one day or two.

**Comment by flavianojs** [comment]
Hi @ltalirz, would you have a look at it now? Cheers.

**Review by ltalirz — APPROVED** [review]
looks great, thanks a lot @flavianojs and happy new year!

**Pull request merged**